realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java

```diff
@@ -307,6 +307,27 @@ private boolean categorizeClassElements() {
                     }
                 }
 
+                if (variableElement.getAnnotation(Required.class) == null) {
+                    // The field doesn't have the @Required annotation
+                    if (!Utils.isPrimitiveType(variableElement) &&
+                            !Utils.isRealmList(variableElement)) {
+                        nullableElements.add(variableElement);
+                    }
+                } else {
+                    // The field has the @Required annotation
+                    if (Utils.isPrimitiveType(variableElement)) {
+                        Utils.error("@Required is not needed for primitive fields - got " + element);
+                    } else if (Utils.isRealmList(variableElement)) {
+                        Utils.error("@Required is not needed for RealmList fields - got " + element);
+                    } else if (Utils.isRealmObject(variableElement)) {
+                        Utils.error("@Required cannot be applied to RealmObject fields - got " + element);
+                    } else {
+                        if (nullableElements.contains(variableElement)) {
```

Comment:
Done. And added an exception in this case (Which i cannot figure out how to trigger it, but anyway, maybe user's strange model definition would trigger this strange path).